File: src/core/parser/parser.go
package parser

import (
	"fmt"
	"quill/src/core/ast"
	"quill/src/core/types"
)

// ─── Parser ──────────────────────────────────────────────────────────────────

type Parser struct {
	tokens  []types.Token
	current int
}

func NewParser(tokens []types.Token) *Parser {
	// Strip semicolons — purely decorative
	var filtered []types.Token
	for _, t := range tokens {
		if t.Value != ";" {
			filtered = append(filtered, t)
		}
	}
	filtered = append(filtered, types.Token{Type: types.EOF, Value: ""})
	return &Parser{tokens: filtered, current: 0}
}

// ─── Position / Error Helpers ──────────────────────────────────────────────

func (p *Parser) pos() int {
	return p.current
}

func (p *Parser) peek() types.Token {
	if p.current >= len(p.tokens) {
		return types.Token{Type: types.EOF, Value: ""}
	}
	return p.tokens[p.current]
}

func (p *Parser) advance() types.Token {
	tok := p.tokens[p.current]
	p.current++
	return tok
}

func (p *Parser) isAtEnd() bool {
	return p.peek().Type == types.EOF
}

func (p *Parser) errf(format string, args ...interface{}) error {
	return fmt.Errorf("token %d: %s", p.pos(), fmt.Sprintf(format, args...))
}

func (p *Parser) consume(expected string) (types.Token, error) {
	if p.peek().Value != expected {
		return types.Token{}, p.errf("expected '%s', got '%s'", expected, p.peek().Value)
	}
	return p.advance(), nil
}

func (p *Parser) peekIs(value string) bool {
	return p.peek().Value == value
}

func (p *Parser) peekTypeIs(t types.TokenType) bool {
	return p.peek().Type == t
}

// ─── Expression Boundary ───────────────────────────────────────────────────

func (p *Parser) canContinueExpression() bool {
	if p.isAtEnd() {
		return false
	}
	switch p.peek().Value {
	case "}", ")", ",", "{", "]", "else", "elif":
		return false
	}
	if p.peek().Type == types.EOF {
		return false
	}
	return true
}

// ─── Program ─────────────────────────────────────────────────────────────────

func (p *Parser) Parse() (ast.Program, error) {
	program := ast.Program{}
	for !p.isAtEnd() {
		if p.peekTypeIs(types.Comment) {
			p.advance()
			continue
		}
		stmt, err := p.parseStatement()
		if err != nil {
			return ast.Program{}, err
		}
		program.Body = append(program.Body, stmt)
	}
	return program, nil
}

// ─── Statements ──────────────────────────────────────────────────────────────

func (p *Parser) parseStatement() (ast.Node, error) {
	tok := p.peek()

	if tok.Type == types.Keyword {
		switch tok.Value {
		case "let", "const":
			return p.parseVariableDeclaration()
		case "printf", "say":
			return p.parsePrintStatement()
		case "func":
			return p.parseFunctionDeclaration()
		case "if", "elif", "else":
			return p.parseConditionalExpression()
		case "while":
			return p.parseLoopExpression()
		case "return":
			return p.parseReturnStatement()
		case "true", "false":
			return p.parseExpressionStatement()
		}
	}

	switch tok.Type {
	case types.Identifier:
		if p.isAssignmentStart() {
			return p.parseAssignment()
		}
		return p.parseExpressionStatement()
	case types.Str, types.Float, types.Number:
		return p.parseExpressionStatement()
	case types.Comment:
		return p.parseComment()
	}

	if tok.Value == "(" {
		return p.parseExpressionStatement()
	}

	return nil, p.errf("unexpected token '%s' (%s) at statement level", tok.Value, tok.Type)
}

func (p *Parser) isAssignmentStart() bool {
	if p.current+1 >= len(p.tokens) {
		return false
	}
	next := p.tokens[p.current+1].Value
	return next == "=" || next == "+=" || next == "-=" || next == "*=" || next == "/=" || next == "%="
}

func (p *Parser) parseExpressionStatement() (ast.Node, error) {
	expr, err := p.parseExpression()
	if err != nil {
		return nil, err
	}
	return ast.ExpressionStatement{Expression: expr}, nil
}

// ─── Variable Declaration ────────────────────────────────────────────────────

func (p *Parser) parseVariableDeclaration() (ast.Node, error) {
	isConst := p.peek().Value == "const"
	p.advance()

	idToken := p.advance()
	if idToken.Type != types.Identifier {
		return nil, p.errf("expected variable name after 'let'/'const', got '%s'", idToken.Value)
	}

	declaredType := ""
	if p.peekIs(":") {
		p.advance()
		typeTok := p.advance()
		declaredType = typeTok.Value
	}

	if _, err := p.consume("="); err != nil {
		return nil, p.errf("expected '=' after variable name, got '%s'", p.peek().Value)
	}

	val, err := p.parseExpression()
	if err != nil {
		return nil, err
	}

	return ast.VariableDeclaration{
		Identifier:     idToken.Value,
		DeclaredType: declaredType,
		Value:          val,
		Mutable:        !isConst, // FIXED: const = not mutable, let = mutable
	}, nil
}

// ─── Assignment ──────────────────────────────────────────────────────────────

func (p *Parser) parseAssignment() (ast.Node, error) {
	idToken := p.advance()
	if idToken.Type != types.Identifier {
		return nil, p.errf("expected identifier, got '%s'", idToken.Value)
	}

	compoundOps := map[string]string{
		"+=": "+",
		"-=": "-",
		"*=": "*",
		"/=": "/",
		"%=": "%",
	}

	if op, ok := compoundOps[p.peek().Value]; ok {
		p.advance()
		val, err := p.parseExpression()
		if err != nil {
			return nil, err
		}
		return ast.Assignment{
			Identifier: idToken.Value,
			Value: ast.BinaryExpression{
				Left:     ast.Identifier{Name: idToken.Value},
				Operator: op,
				Right:    val,
			},
		}, nil
	}

	if _, err := p.consume("="); err != nil {
		return nil, p.errf("expected '=' after identifier '%s', got '%s'", idToken.Value, p.peek().Value)
	}

	val, err := p.parseExpression()
	if err != nil {
		return nil, err
	}

	return ast.Assignment{
		Identifier: idToken.Value,
		Value:      val,
	}, nil
}

// ─── Print Statement ───────────────────────────────────────────────────────────

func (p *Parser) parsePrintStatement() (ast.Node, error) {
	p.advance()
	val, err := p.parseExpression()
	if err != nil {
		return nil, err
	}
	return ast.PrintStatement{Expression: val}, nil
}

// ─── Return Statement ──────────────────────────────────────────────────────────

func (p *Parser) parseReturnStatement() (ast.Node, error) {
	p.advance()
	if p.peekIs("}") || p.isAtEnd() {
		return ast.ReturnStatement{Value: nil}, nil
	}
	val, err := p.parseExpression()
	if err != nil {
		return nil, err
	}
	return ast.ReturnStatement{Value: val}, nil
}

// ─── Comment ─────────────────────────────────────────────────────────────────

func (p *Parser) parseComment() (ast.Node, error) {
	tok := p.advance()
	return ast.Comment{Value: tok.Value}, nil
}

// ─── Function Declaration ────────────────────────────────────────────────────

func (p *Parser) parseFunctionDeclaration() (ast.Node, error) {
	p.advance()

	nameTok := p.advance()
	if nameTok.Type != types.Identifier {
		return nil, p.errf("expected function name after 'func', got '%s'", nameTok.Value)
	}

	if _, err := p.consume("("); err != nil {
		return nil, err
	}

	var params []ast.Param
	for !p.isAtEnd() && !p.peekIs(")") {
		paramTok := p.advance()
		if paramTok.Type != types.Identifier {
			return nil, p.errf("expected parameter name, got '%s'", paramTok.Value)
		}
		paramType := ""
		if p.peekIs(":") {
			p.advance()
			paramType = p.advance().Value
		}
		params = append(params, ast.Param{Name: paramTok.Value, Type: paramType})
		if p.peekIs(",") {
			p.advance()
		}
	}

	if _, err := p.consume(")"); err != nil {
		return nil, err
	}

	returnType := ""
	if p.peekIs(":") {
		p.advance()
		returnType = p.advance().Value
	}

	if _, err := p.consume("{"); err != nil {
		return nil, err
	}

	var body []ast.Node
	for !p.isAtEnd() && !p.peekIs("}") {
		stmt, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		body = append(body, stmt)
	}

	if _, err := p.consume("}"); err != nil {
		return nil, err
	}

	return ast.Function{
		Name:       nameTok.Value,
		Params:     params,
		ReturnType: returnType,
		Body:       body,
	}, nil
}

// ─── Loop Expression ───────────────────────────────────────────────────────────

func (p *Parser) parseLoopExpression() (ast.Node, error) {
	p.advance()

	test, err := p.parseExpression()
	if err != nil {
		return nil, err
	}

	if _, err := p.consume("{"); err != nil {
		return nil, err
	}

	var body []ast.Node
	for !p.isAtEnd() && !p.peekIs("}") {
		stmt, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		body = append(body, stmt)
	}

	if _, err := p.consume("}"); err != nil {
		return nil, err
	}

	return ast.WhileLoop{Condition: test, Body: body}, nil
}

// ─── Conditional Expression ────────────────────────────────────────────────────

func (p *Parser) parseConditionalExpression() (ast.Node, error) {
	p.advance()

	test, err := p.parseExpression()
	if err != nil {
		return nil, err
	}

	if _, err := p.consume("{"); err != nil {
		return nil, err
	}

	var consequent []ast.Node
	for !p.isAtEnd() && !p.peekIs("}") {
		stmt, err := p.parseStatement()
		if err != nil {
			return nil, err
		}
		consequent = append(consequent, stmt)
	}

	if _, err := p.consume("}"); err != nil {
		return nil, err
	}

	var alternate []ast.Node
	if p.peekIs("else") {
		p.advance()
		if p.peekIs("{") {
			// else { ... }
			p.advance()
			for !p.isAtEnd() && !p.peekIs("}") {
				stmt, err := p.parseStatement()
				if err != nil {
					return nil, err
				}
				alternate = append(alternate, stmt)
			}
			if _, err := p.consume("}"); err != nil {
				return nil, err
			}
		} else if p.peekIs("if") || p.peekIs("elif") {
			// else if / else elif
			nested, err := p.parseConditionalExpression()
			if err != nil {
				return nil, err
			}
			alternate = []ast.Node{nested}
		}
	} else if p.peekIs("elif") {
		// elif without preceding else
		nested, err := p.parseConditionalExpression()
		if err != nil {
			return nil, err
		}
		alternate = []ast.Node{nested}
	}

	return ast.IfStatement{
		Condition:  test,
		Consequent: consequent,
		Alternate:  alternate,
	}, nil
}

// ─── Expressions ───────────────────────────────────────────────────────────────

var binaryOps = map[string]bool{
	"+": true, "-": true, "*": true, "/": true,
	"==": true, "!=": true, ">": true, "<": true,
	">=": true, "<=": true, "&&": true, "||": true,
}

func (p *Parser) isBinaryOperator(tok types.Token) bool {
	return tok.Type == types.Operator && binaryOps[tok.Value]
}

func (p *Parser) parseExpression() (ast.Node, error) {
	return p.parsePrecedence(0)
}

var precedence = map[string]int{
	"||": 1,
	"&&": 2,
	"==": 3, "!=": 3,
	"<": 4, ">": 4, "<=": 4, ">=": 4,
	"+": 5, "-": 5,
	"*": 6, "/": 6, "%": 6,
}

func (p *Parser) parsePrecedence(minPrec int) (ast.Node, error) {
	left, err := p.parsePrimary()
	if err != nil {
		return nil, err
	}

	// Postfix
	for p.canContinueExpression() {
		tok := p.peek()
		if tok.Value == "[" {
			p.advance()
			index, err := p.parseExpression()
			if err != nil {
				return nil, err
			}
			if _, err := p.consume("]"); err != nil {
				return nil, err
			}
			left = ast.IndexExpression{Object: left, Index: index}
		} else if tok.Type == types.Incrementation {
			op := p.advance().Value
			name := ""
			switch n := left.(type) {
			case ast.Identifier:
				name = n.Name
			}
			if op == "++" {
				left = ast.Increment{Identifier: name}
			} else {
				left = ast.Decrement{Identifier: name}
			}
		} else {
			break
		}
	}

	// Binary with precedence climbing
	for p.canContinueExpression() {
		tok := p.peek()
		prec, ok := precedence[tok.Value]
		if !ok || prec < minPrec {
			break
		}
		if !p.isBinaryOperator(tok) {
			break
		}
		op := p.advance().Value
		right, err := p.parsePrecedence(prec + 1)
		if err != nil {
			return nil, err
		}
		left = ast.BinaryExpression{Left: left, Operator: op, Right: right}
	}

	return left, nil
}

func (p *Parser) parsePrimary() (ast.Node, error) {
	tok := p.peek()

	// Unary operators — ONLY when token is actually an Operator
	if tok.Type == types.Operator && (tok.Value == "!" || tok.Value == "-") {
		op := p.advance().Value
		operand, err := p.parsePrimary()
		if err != nil {
			return nil, err
		}
		return ast.UnaryExpression{Operator: op, Operand: operand}, nil
	}

	// Parenthesized expression
	if tok.Value == "(" {
		p.advance()
		expr, err := p.parseExpression()
		if err != nil {
			return nil, err
		}
		if _, err := p.consume(")"); err != nil {
			return nil, err
		}
		return expr, nil
	}

	consumed := p.advance()

	switch consumed.Type {
	case types.Number:
		var val int64
		fmt.Sscanf(consumed.Value, "%d", &val)
		return ast.LiteralInt{Value: val}, nil

	case types.Float:
		var val float64
		fmt.Sscanf(consumed.Value, "%f", &val)
		return ast.LiteralFloat{Value: val}, nil

	case types.Str:
		return ast.LiteralString{Value: consumed.Value}, nil

	case types.Keyword:
		switch consumed.Value {
		case "true":
			return ast.LiteralBool{Value: true}, nil
		case "false":
			return ast.LiteralBool{Value: false}, nil
		}

	case types.Identifier:
		// Function call
		if p.peekIs("(") {
			p.advance()
			var args []ast.Node
			for !p.isAtEnd() && !p.peekIs(")") {
				arg, err := p.parseExpression()
				if err != nil {
					return nil, err
				}
				args = append(args, arg)
				if p.peekIs(",") {
					p.advance()
				}
			}
			if _, err := p.consume(")"); err != nil {
				return nil, err
			}
			return ast.CallExpression{Callee: consumed.Value, Arguments: args}, nil
		}
		return ast.Identifier{Name: consumed.Value}, nil
	}

	return nil, p.errf("expected expression, got '%s'", consumed.Value)
}

File: src/core/parser/parser.cpp
#pragma once

#include <algorithm>
#include <string>
#include <vector>

#include "../ast/ast.cpp"
#include "../types/types.cpp"

class Parser
{
public:
    explicit Parser(std::vector<Token> tokens) : tokens_(std::move(tokens)), index_(0) {}

    Program parse()
    {
        Program program;
        while (!isAtEnd())
        {
            if (peek().type == TokenType::Comment)
            {
                advance();
                continue;
            }
            if (peek().value == ";")
            {
                advance();
                continue;
            }

            Node *stmt = parseStatement();
            if (stmt)
            {
                program.body.push_back(stmt);
            }
        }
        return program;
    }

private:
    bool isAtEnd() const
    {
        return peek().type == TokenType::EOF_T;
    }

    Token peek() const
    {
        if (index_ >= tokens_.size())
            return Token{TokenType::EOF_T, ""};
        return tokens_[index_];
    }

    Token advance()
    {
        if (index_ >= tokens_.size())
            return Token{TokenType::EOF_T, ""};
        return tokens_[index_++];
    }

    bool isBinaryOperator(const Token &token) const
    {
        if (token.type != TokenType::Operator)
            return false;
        static const std::vector<std::string> ops = {"+", "-", "*", "/", "%", "==", "!=", ">", "<", ">=", "<=", "&&", "||"};
        return std::find(ops.begin(), ops.end(), token.value) != ops.end();
    }

    Node *parseStatement()
    {
        // Comments were only ever skipped by the top-level parse() loop,
        // so one written inside a function/if/while body would fall
        // through every check below and hit the catch-all -- which
        // consumes the comment's own token, not the comment, and hands
        // back an empty Identifier as if it were real code. Skipping
        // here instead means every nested body-parsing loop (which all
        // call parseStatement in a loop and already handle a null
        // return the same way ';' is handled) gets this for free.
        if (peek().type == TokenType::Comment)
        {
            advance();
            return nullptr;
        }

        if (peek().value == ";")
        {
            advance();
            return nullptr;
        }

        if (peek().value == "let" || peek().value == "const" || peek().value == "mut")
        {
            return parseVariableDeclaration();
        }

        if (peek().value == "func" || peek().value == "fn")
        {
            return parseFunctionDeclaration();
        }

        if (peek().value == "extern")
        {
            advance(); // "extern"
            Node *node = parseFunctionDeclaration();
            if (auto *fn = dynamic_cast<Function *>(node))
            {
                fn->isExtern = true;
            }
            return node;
        }

        if (peek().value == "printf" || peek().value == "say")
        {
            return parsePrintStatement();
        }

        if (peek().value == "return")
        {
            return parseReturnStatement();
        }

        if (peek().value == "if")
        {
            return parseIfStatement();
        }

        if (peek().value == "while")
        {
            return parseWhileLoop();
        }

        if (peek().value == "import")
        {
            return parseImportStatement();
        }

        if (peek().type == TokenType::Identifier)
        {
            return parseIdentifierStatement();
        }

        if (peek().type == TokenType::Number || peek().type == TokenType::Double || peek().type == TokenType::String)
        {
            return parseLiteralStatement();
        }

        if (peek().value == "true" || peek().value == "false")
        {
            return parseLiteralStatement();
        }

        ExpressionStatement *node = new ExpressionStatement();
        node->expression = parseExpression();
        return node;
    }

    Node *parseVariableDeclaration()
    {
        std::string kind = advance().value;
        std::string name = advance().value;

        std::string type = "";
        if (peek().value == ":")
        {
            advance();
            type = advance().value;

            // Array type: `number[200]` -- a fixed size, given as a
            // literal, folded straight into the type string ("number[200]")
            // so nothing else needs a new AST field for it.
            if (peek().value == "[")
            {
                advance();
                std::string sizeTok = advance().value;
                type += "[" + sizeTok + "]";
                if (peek().value == "]")
                {
                    advance();
                }
            }
        }

        // Arrays are declared with a fixed size and no initializer --
        // `let nums: number[5];` -- so only parse a value if `=` was
        // actually written.
        Node *value = nullptr;
        if (peek().value == "=")
        {
            advance();
            value = parseExpression();
        }

        VariableDeclaration *node = new VariableDeclaration();
        node->identifier = name;
        node->declaredType = type;
        node->value = value;
        (void)kind;
        return node;
    }

    Node *parseFunctionDeclaration()
    {
        advance(); // func
        Function *node = new Function();

        Token nameTok = advance();
        node->name = nameTok.value;

        if (peek().value == "(")
        {
            advance();
            while (!isAtEnd() && peek().value != ")")
            {
                Token paramName = advance();
                Param p;
                p.name = paramName.value;
                if (peek().value == ":")
                {
                    advance();
                    p.type = advance().value;
                }
                node->params.push_back(p);
                if (peek().value == ",")
                {
                    advance();
                }
            }
            if (peek().value == ")")
            {
                advance();
            }
        }

        if (peek().value == ":")
        {
            advance();
            node->returnType = advance().value;
        }

        if (peek().value == "{")
        {
            advance();
            while (!isAtEnd() && peek().value != "}")
            {
                Node *stmt = parseStatement();
                if (stmt)
                {
                    node->body.push_back(stmt);
                }
            }
            if (peek().value == "}")
            {
                advance();
            }
        }

        return node;
    }

    Node *parsePrintStatement()
    {
        advance();
        PrintStatement *node = new PrintStatement();
        node->expression = parseExpression();
        return node;
    }

    Node *parseReturnStatement()
    {
        advance(); // "return"

        // A bare `return;` with no value is valid (early-exit from a
        // void function). Only parse an expression if the next token
        // could actually start one -- otherwise, since ';' produces no
        // token at all (the lexer discards it), parseExpression() would
        // fall through to its catch-all and consume whatever comes
        // next instead (e.g. a block's closing '}'), desynchronizing
        // the rest of the parse.
        bool hasValue = peek().type == TokenType::Identifier ||
                        peek().type == TokenType::Number ||
                        peek().type == TokenType::Double ||
                        peek().type == TokenType::String ||
                        peek().value == "true" || peek().value == "false" ||
                        peek().value == "(" ||
                        (peek().type == TokenType::Operator && (peek().value == "!" || peek().value == "-"));

        ReturnStatement *node = new ReturnStatement();
        node->value = hasValue ? parseExpression() : nullptr;
        return node;
    }

    Node *parseIfStatement()
    {
        advance();
        IfStatement *node = new IfStatement();
        node->condition = parseExpression();

        if (peek().value == "{")
        {
            advance();
            while (!isAtEnd() && peek().value != "}")
            {
                Node *stmt = parseStatement();
                if (stmt)
                {
                    node->consequent.push_back(stmt);
                }
            }
            if (peek().value == "}")
            {
                advance();
            }
        }

        if (peek().value == "else")
        {
            advance();
            if (peek().value == "if")
            {
                Node *nested = parseIfStatement();
                node->alternate.push_back(nested);
            }
            else if (peek().value == "{")
            {
                advance();
                while (!isAtEnd() && peek().value != "}")
                {
                    Node *stmt = parseStatement();
                    if (stmt)
                    {
                        node->alternate.push_back(stmt);
                    }
                }
                if (peek().value == "}")
                {
                    advance();
                }
            }
        }

        return node;
    }

    Node *parseWhileLoop()
    {
        advance();
        WhileLoop *node = new WhileLoop();
        node->condition = parseExpression();

        if (peek().value == "{")
        {
            advance();
            while (!isAtEnd() && peek().value != "}")
            {
                Node *stmt = parseStatement();
                if (stmt)
                {
                    node->body.push_back(stmt);
                }
            }
            if (peek().value == "}")
            {
                advance();
            }
        }

        return node;
    }

    Node *parseImportStatement()
    {
        advance(); // "import"

        ImportStatement *node = new ImportStatement();
        if (peek().type == TokenType::String)
        {
            node->path = advance().value;
        }
        return node;
    }

    Node *parseIdentifierStatement()
    {
        std::string name = advance().value;

        if (peek().value == "[")
        {
            advance();
            Node *indexExpr = parseExpression();
            if (peek().value == "]")
            {
                advance();
            }

            if (peek().value == "=")
            {
                advance();
                IndexAssignment *node = new IndexAssignment();
                Identifier *obj = new Identifier();
                obj->name = name;
                node->object = obj;
                node->index = indexExpr;
                node->value = parseExpression();
                return node;
            }

            // `name[i]` used as a bare statement (rare, but falls back
            // safely rather than dropping the tokens).
            IndexExpression *idx = new IndexExpression();
            Identifier *obj = new Identifier();
            obj->name = name;
            idx->object = obj;
            idx->index = indexExpr;
            return idx;
        }

        if (peek().value == "=")
        {
            advance();
            Assignment *node = new Assignment();
            node->identifier = name;
            node->value = parseExpression();
            return node;
        }

        if (peek().value == "(")
        {
            advance();
            CallExpression *node = new CallExpression();
            node->callee = name;
            while (!isAtEnd() && peek().value != ")")
            {
                node->arguments.push_back(parseExpression());
                if (peek().value == ",")
                {
                    advance();
                }
            }
            if (peek().value == ")")
            {
                advance();
            }
            return node;
        }

        if (peek().value == "++" || peek().value == "--")
        {
            std::string op = advance().value;
            if (op == "++")
            {
                auto *node = new Increment();
                node->identifier = name;
                return node;
            }
            else
            {
                auto *node = new Decrement();
                node->identifier = name;
                return node;
            }
        }

        Identifier *node = new Identifier();
        node->name = name;
        return node;
    }

    Node *parseLiteralStatement()
    {
        Token tok = advance();

        if (tok.type == TokenType::Number)
        {
            LiteralInt *val = new LiteralInt();
            val->value = std::stoll(tok.value);
            return val;
        }

        if (tok.type == TokenType::Double)
        {
            LiteralDoudle *val = new LiteralDoudle();
            val->value = std::stod(tok.value);
            return val;
        }

        if (tok.type == TokenType::String)
        {
            LiteralString *val = new LiteralString();
            val->value = tok.value;
            return val;
        }

        if (tok.value == "true")
        {
            LiteralBool *val = new LiteralBool();
            val->value = true;
            return val;
        }

        if (tok.value == "false")
        {
            LiteralBool *val = new LiteralBool();
            val->value = false;
            return val;
        }

        return new Identifier();
    }

    Node *parseExpression()
    {
        return parsePrecedence(0);
    }

    Node *parsePrecedence(int minPrecedence)
    {
        Node *left = parsePrimary();

        while (!isAtEnd() && peek().type == TokenType::Operator && isBinaryOperator(peek()))
        {
            std::string op = peek().value;
            int precedence = precedenceOf(op);
            if (precedence < minPrecedence)
            {
                break;
            }
            advance();

            Node *previousLeft = left;
            Node *right = parsePrecedence(precedence + 1);

            BinaryExpression *expr = new BinaryExpression();
            expr->op = op;
            expr->left = previousLeft;
            expr->right = right;
            left = expr;
        }

        return left;
    }

    int precedenceOf(const std::string &op) const
    {
        if (op == "||")
            return 1;
        if (op == "&&")
            return 2;
        if (op == "==" || op == "!=")
            return 3;
        if (op == "<" || op == ">" || op == "<=" || op == ">=")
            return 4;
        if (op == "+" || op == "-")
            return 5;
        if (op == "*" || op == "/" || op == "%")
            return 6;
        return -1;
    }

    Node *parsePrimary()
    {
        if (peek().type == TokenType::Identifier)
        {
            std::string name = advance().value;
            if (peek().value == "(")
            {
                advance();
                CallExpression *node = new CallExpression();
                node->callee = name;
                while (!isAtEnd() && peek().value != ")")
                {
                    node->arguments.push_back(parseExpression());
                    if (peek().value == ",")
                    {
                        advance();
                    }
                }
                if (peek().value == ")")
                {
                    advance();
                }
                return node;
            }

            if (peek().value == "[")
            {
                advance();
                IndexExpression *idx = new IndexExpression();
                Identifier *obj = new Identifier();
                obj->name = name;
                idx->object = obj;
                idx->index = parseExpression();
                if (peek().value == "]")
                {
                    advance();
                }
                return idx;
            }

            Identifier *node = new Identifier();
            node->name = name;
            return node;
        }

        if (peek().type == TokenType::Number)
        {
            LiteralInt *val = new LiteralInt();
            val->value = std::stoll(advance().value);
            return val;
        }

        if (peek().type == TokenType::Double)
        {
            LiteralDoudle *val = new LiteralDoudle();
            val->value = std::stod(advance().value);
            return val;
        }

        if (peek().type == TokenType::String)
        {
            LiteralString *val = new LiteralString();
            val->value = advance().value;
            return val;
        }

        if (peek().value == "true" || peek().value == "false")
        {
            LiteralBool *val = new LiteralBool();
            val->value = (advance().value == "true");
            return val;
        }

        if (peek().value == "(")
        {
            advance();
            Node *inner = parseExpression();
            if (peek().value == ")")
            {
                advance();
            }
            return inner;
        }

        if (peek().type == TokenType::Operator && (peek().value == "!" || peek().value == "-"))
        {
            UnaryExpression *node = new UnaryExpression();
            node->op = advance().value;
            node->operand = parsePrimary();
            return node;
        }

        Token t = advance();
        (void)t;
        return new Identifier();
    }

    std::vector<Token> tokens_;
    size_t index_;
};